Public/Src/Engine/Processes/ExternalSandboxedProcess.cs

```diff
@@ -0,0 +1,157 @@
+        /// the deserialization result of sandboxed process result. This output directory can also contain the dump of the external process
+        /// when it gets killed.
+        /// </remarks>
+        protected string GetOutputDirectory() => Path.GetDirectoryName(SandboxedProcessInfo.FileStorage.GetFileName(SandboxedProcessFile.StandardOutput));
+
+        /// <summary>
+        /// Gets the standard output path for the external executor; not the detoured process.
+        /// </summary>
+        protected string GetStdOutPath(string hint) => Path.Combine(GetOutputDirectory(), $"{hint ?? string.Empty}-Pip{SandboxedProcessInfo.PipSemiStableHash:X16}.out");
+
+        /// <summary>
+        /// Gets the standard error path for the external executor; not the detoured process.
+        /// </summary>
+        protected string GetStdErrPath(string hint) => Path.Combine(GetOutputDirectory(), $"{hint ?? string.Empty}-Pip{SandboxedProcessInfo.PipSemiStableHash:X16}.err");
+
+        /// <summary>
+        /// Standard output for the external executor.
+        /// </summary>
+        public abstract string StdOut { get; }
+
+        /// <summary>
+        /// Standard error for the external executor.
+        /// </summary>
+        public abstract string StdErr { get; }
+
+        /// <summary>
+        /// Gets the exit code of the external executor.
+        /// </summary>
+        public abstract int? ExitCode { get; }
+
+        /// <summary>
+        /// Serializes sandboxed process info to file.
+        /// </summary>
+        protected void SerializeSandboxedProcessInfoToFile()
+        {
+            string file = GetSandboxedProcessInfoFile();
+            FileUtilities.CreateDirectory(Path.GetDirectoryName(file));
+
+            try
+            {
+                using (FileStream stream = File.OpenWrite(file))
+                {
+                    SandboxedProcessInfo.Serialize(stream);
+                }
+            }
+            catch (IOException ioException)
+            {
+                ThrowBuildXLException($"Failed to serialize sandboxed process info '{file}'", ioException);
+            }
+        }
+
+        /// <summary>
+        /// Deserializes sandboxed process result from file.
+        /// </summary>
+        /// <returns></returns>
+        protected SandboxedProcessResult DeserializeSandboxedProcessResultFromFile()
+        {
+            string file = GetSandboxedProcessResultsFile();
+
+            try
+            {
+                using (FileStream stream = File.OpenRead(file))
+                {
+                    return SandboxedProcessResult.Deserialize(stream);
+                }
+            }
+            catch (IOException ioException)
+            {
+                ThrowBuildXLException($"Failed to deserialize sandboxed process result '{file}'", ioException);
+            }
+
+            return null;

```

Comment:
nit: maybe move inside the `catch` (to make it clearer that this is unreachable)